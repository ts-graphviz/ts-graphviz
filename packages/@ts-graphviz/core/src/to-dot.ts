import { DotObjectModel } from '@ts-graphviz/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '@ts-graphviz/ast';

/**
 * This interface provides options for converting a model to DOT.
 * @group Convert Model to DOT
 */
export interface ToDotOptions {
  /**
   * Options for converting the model to DOT.
   */
  convert?: ConvertFromModelOptions;
  /**
   * Options for printing DOT.
   */
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Subgraph}, {@link Node}, and {@link Edge}
 * @param options Optional options for the conversion.
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
