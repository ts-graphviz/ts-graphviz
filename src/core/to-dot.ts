import { DotObjectModel } from '#lib/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '#lib/ast';

/**
 * @group Convert Model to DOT
 * @alpha
 */
export interface ToDotOptions {
  convert?: ConvertFromModelOptions;
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Node}, and {@link Edge}
 * @param options
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
