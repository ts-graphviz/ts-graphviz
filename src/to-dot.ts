import { DotObjectModel } from './common/index.js';
import { stringify, fromModel, PrintOptions } from './ast/index.js';
import { ConvertOptions } from './ast/index.js';

export interface ToDotOptions {
  convert?: ConvertOptions;
  print?: PrintOptions;
}

export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
