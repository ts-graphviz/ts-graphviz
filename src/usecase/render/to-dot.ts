import { Renderer, Dot } from './renderer.js';

export function toDot(object: Dot): string {
  const renderer = new Renderer();
  return renderer.render(object);
}
