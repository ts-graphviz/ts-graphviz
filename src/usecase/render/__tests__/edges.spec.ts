import { Graph, Node, Edge } from '../../../model/index.js';
import { toDot } from '../to-dot.js';

describe('Edge rendering', () => {
  let edge: Edge;

  const [node1, node2, node3, node4] = [...Array(4)].map((_, i) => new Node(`node${i + 1}`));

  beforeEach(() => {
    edge = new Edge([node1, node2]);
  });

  describe.each([
    ['Digraph context', (): Graph => new Graph(true)],
    ['Graph context', (): Graph => new Graph(false)],
  ])('%s', (_, create) => {
    let root: Graph;
    beforeEach(() => {
      root = create();
      root.addEdge(edge);
    });
    it('simple edge', () => {
      expect(toDot(root)).toMatchSnapshot();
    });

    describe('edge with comment', () => {
      test('single line comment', () => {
        edge.comment = 'this is comment.';
        expect(toDot(root)).toMatchSnapshot();
      });

      test('multi line comment', () => {
        edge.comment = 'this is comment.\nsecond line.';
        expect(toDot(root)).toMatchSnapshot();
      });
    });

    it('has some attributes', () => {
      edge.attributes.set('label', 'this is test');
      edge.attributes.set('color', 'red');
      expect(toDot(root)).toMatchSnapshot();
    });

    describe('custom edge', () => {
      beforeEach(() => {
        root = create();
      });

      it('node group', () => {
        edge = new Edge([node1, [node2, node3], node4]);
        root.addEdge(edge);
        expect(toDot(root)).toMatchSnapshot();
      });

      it('3 nodes', () => {
        edge = new Edge([node1, node2, node3]);
        root.addEdge(edge);
        expect(toDot(root)).toMatchSnapshot();
      });

      it('3 nodes, but many args', () => {
        edge = new Edge([node1, node2, node3, node1, node2, node3, node1, node2, node3]);
        root.addEdge(edge);
        expect(toDot(root)).toMatchSnapshot();
      });
    });
  });
});
