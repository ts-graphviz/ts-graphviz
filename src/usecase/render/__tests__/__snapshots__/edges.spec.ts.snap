// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edge rendering Digraph context custom edge 3 nodes 1`] = `
"digraph {
  \\"node1\\" -> \\"node2\\" -> \\"node3\\";
}"
`;

exports[`Edge rendering Digraph context custom edge 3 nodes, but many args 1`] = `
"digraph {
  \\"node1\\" -> \\"node2\\" -> \\"node3\\" -> \\"node1\\" -> \\"node2\\" -> \\"node3\\" -> \\"node1\\" -> \\"node2\\" -> \\"node3\\";
}"
`;

exports[`Edge rendering Digraph context custom edge node group 1`] = `
"digraph {
  \\"node1\\" -> {\\"node2\\" \\"node3\\"} -> \\"node4\\";
}"
`;

exports[`Edge rendering Digraph context edge with comment multi line comment 1`] = `
"digraph {
  // this is comment.
  // second line.
  \\"node1\\" -> \\"node2\\";
}"
`;

exports[`Edge rendering Digraph context edge with comment single line comment 1`] = `
"digraph {
  // this is comment.
  \\"node1\\" -> \\"node2\\";
}"
`;

exports[`Edge rendering Digraph context has some attributes 1`] = `
"digraph {
  \\"node1\\" -> \\"node2\\" [
    label = \\"this is test\\",
    color = \\"red\\",
  ];
}"
`;

exports[`Edge rendering Digraph context simple edge 1`] = `
"digraph {
  \\"node1\\" -> \\"node2\\";
}"
`;

exports[`Edge rendering Graph context custom edge 3 nodes 1`] = `
"graph {
  \\"node1\\" -- \\"node2\\" -- \\"node3\\";
}"
`;

exports[`Edge rendering Graph context custom edge 3 nodes, but many args 1`] = `
"graph {
  \\"node1\\" -- \\"node2\\" -- \\"node3\\" -- \\"node1\\" -- \\"node2\\" -- \\"node3\\" -- \\"node1\\" -- \\"node2\\" -- \\"node3\\";
}"
`;

exports[`Edge rendering Graph context custom edge node group 1`] = `
"graph {
  \\"node1\\" -- {\\"node2\\" \\"node3\\"} -- \\"node4\\";
}"
`;

exports[`Edge rendering Graph context edge with comment multi line comment 1`] = `
"graph {
  // this is comment.
  // second line.
  \\"node1\\" -- \\"node2\\";
}"
`;

exports[`Edge rendering Graph context edge with comment single line comment 1`] = `
"graph {
  // this is comment.
  \\"node1\\" -- \\"node2\\";
}"
`;

exports[`Edge rendering Graph context has some attributes 1`] = `
"graph {
  \\"node1\\" -- \\"node2\\" [
    label = \\"this is test\\",
    color = \\"red\\",
  ];
}"
`;

exports[`Edge rendering Graph context simple edge 1`] = `
"graph {
  \\"node1\\" -- \\"node2\\";
}"
`;
